### PR TITLE
Add class for the Matrix API V1

### LINKED ISSUE #212: Implement Matrix API v1
Reference: https://www.mapbox.com/api-documentation/#matrix.

### PR DESCRIPTION
**Pull request by sgillies**
Resolves #212

### CODE REVIEW

File: CHANGES [hunk 1]
```diff
@@ -1,7 +1,9 @@
-Next (YYYY-MM-DD)
------------------
+0.15.0 (2018-01-10)
```
Comment (sgillies) [inline CHANGES]: Or thereabouts.

File: CHANGES [hunk 2]
```diff
@@ -1,7 +1,9 @@
...
+-------------------
 - Add api_name and api_version class attributes (#207).
-
+- Add DirectionsMatrix class for access to the Matrix API V1 (#212).
```
Comment (sgillies) [inline CHANGES]: @perrygeo what do you think about the name? The Web API has `directions-matrix` in its endpoint URL and I followed suit.

File: docs/matrix.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+zeros.
+
+```python
+>>> from pprint import pprint
+>>> pprint(response.json()['durations'])
+[[0, ..., ...], [..., 0, ...], [..., ..., 0]]
+
+```
+
+See ``import mapbox; help(mapbox.Distance)`` for more detailed usage.
```
Comment (sgillies) [inline docs/matrix.md]: TODO: replace `Distance` with `DirectionsMatrix`.

You can see that this is copied from distance.md.

File: mapbox/encoding.py [hunk 4]
```diff
@@ -24,7 +24,10 @@ def read_points(features):
     """
     for feature in features:
 
-        if hasattr(feature, '__geo_interface__'):
+        if isinstance(feature, (tuple, list)) and len(feature) == 2:
+            yield feature
```
Comment (sgillies) [inline mapbox/encoding.py]: We can now use sequences of `(lng, lat)`.

File: mapbox/encoding.py [hunk 5]
```diff
@@ -42,6 +45,11 @@ def read_points(features):
             for pt in _geom_points(geom):
                 yield pt
 
+        elif 'coordinates' in feature:
+            geom = feature
+            for pt in _geom_points(geom):
+                yield pt
```
Comment (sgillies) [inline mapbox/encoding.py]: And geometries.

File: mapbox/encoding.py [hunk 6]
```diff
@@ -53,7 +61,7 @@ def encode_waypoints(features, min_limit=None, max_limit=None, precision=6):
     return a string encoded in waypoint-style used by certain mapbox APIs
     ("lon,lat" pairs separated by ";")
     """
-    coords = ['{lon:.{p}f},{lat:.{p}f}'.format(lon=lon, lat=lat, p=precision)
+    coords = ['{lon},{lat}'.format(lon=round(lon, precision), lat=round(lat, precision))
```
Comment (sgillies) [inline mapbox/encoding.py]: Putting the waypoints in the URL as the Matrix API does (which I'm 👎  on to be honest) exposes some bugginess in printing with precision. Rounding and then printing works better and is backwards compatible.

File: mapbox/services/matrix.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    @property
+    def baseuri(self):
+        return 'https://{0}/{1}/{2}'.format(
+            self.host, self.api_name, self.api_version)
+
+    def _validate_profile(self, profile):
+        parts = re.split(r'[\/\.]', profile)
+        if len(parts) == 1:
+            parts = ['mapbox'] + parts
+        profile = '/'.join(parts)
```
Comment (sgillies) [inline mapbox/services/matrix.py]: We've got some inconsistency in the APIs: `mapbox.driving` in some places and `mapbox/driving` in the new API. To help users, we can support each, as well as upgrading bare `driving` to `mapbox/driving`.

File: tests/test_matrix.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    assert serv.api_version == 'v1'
+
+
+def test_profile_invalid():
+    """'jetpack' is not a valid profile."""
+    with pytest.raises(ValueError):
+        mapbox.DirectionsMatrix(access_token='pk.test')._validate_profile('jetpack')
+
+
+@pytest.mark.parametrize('profile', ['mapbox/driving', 'mapbox/cycling', 'mapbox/walking'])
```
Comment (sgillies) [inline tests/test_matrix.py]: Parametrize for the win.

File: tests/test_matrix.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+def test_profile_valid(profile):
+    """Profiles are valid"""
+    assert profile == mapbox.DirectionsMatrix(
+        access_token='pk.test')._validate_profile(profile)
+
+
+@responses.activate
+@pytest.mark.parametrize('waypoints', [points, [p['geometry'] for p in points], [p['geometry']['coordinates'] for p in points]])
```
Comment (sgillies) [inline tests/test_matrix.py]: Parametrize for the win.

File: mapbox/services/matrix.py [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+        if profile not in self.valid_profiles:
+            raise InvalidProfileError(
+                "{0} is not a valid profile".format(profile))
+        return profile
+
+    def matrix(self, features, profile='mapbox/driving'):
```
Comment (sgillies) [inline mapbox/services/matrix.py]: This class has only one method. We could have used `get()` or even `__call__()` but I prefer a more significant name.

File: docs/matrix.md [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+
+`DirectionsMatrix` methods return an instance of
+[`requests.Response`](http://docs.python-requests.org/en/latest/api/#requests.Response).
+If the response is successful, the `json()` method returns Python data parsed
+directly from the API.
+
+## Usage
+
+If you need to optimize travel between several waypoints, you can use the
+Matrix API to create a "Distance Matrix" showing travel times between all
```
Comment (perrygeo) [inline docs/matrix.md]: We should use "Directions Matrix" for consistency

File: docs/matrix.md [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+```python
+>>> response = service.matrix([portland, bend, corvallis], 'driving')
+>>> response.status_code
+200
+>>> response.headers['Content-Type']
+'application/json; charset=utf-8'
+
+```
+
+And the response JSON contains a durations matrix, a 2-D list with travel
```
Comment (perrygeo) [inline docs/matrix.md]: s/durations matrix/Directions matrix/

File: mapbox/services/matrix.py (hunk 7 above)
Comment (perrygeo) [reply mapbox/services/matrix.py]: I took a slightly different approach in the Direction v5 PR: https://github.com/mapbox/mapbox-sdk-py/blob/ee78696f1924ba48b23f1d951f3f1213d849a357/mapbox/services/directions.py#L32-L45

Just noting that there is some overlap and we should consider pulling into a common function. I'd prefer a warning if using the old `.` style.

File: mapbox/services/matrix.py (hunk 10 above)
Comment (perrygeo) [reply mapbox/services/matrix.py]: This is the case with all of our service classes - agree that the explicit name is better but _if_ we wanted to standardize the method name or make the class callable, we should do so across the board before the 1.0 release.

File: CHANGES (hunk 2 above)
Comment (perrygeo) [reply CHANGES]: Directions Matrix sounds good to me. There are a few cases in the docs that we should change for consistency.

**Comment by sgillies** [comment]
@perrygeo I'm validating profiles exactly as we do in the directions v5 branch. Your other requests are resolved, too. This is ready to go.

File: tests/test_matrix.py [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+    """Profiles are valid"""
+    assert profile == DirectionsMatrix(
+        access_token='pk.test')._validate_profile(profile)
+
+
+@pytest.mark.parametrize('profile', ['mapbox.driving', 'mapbox.cycling', 'mapbox.walking'])
+def test_deprecated_profile(profile):
+    """Profiles are deprecated"""
+    service = DirectionsMatrix()
+    with pytest.warns(MapboxDeprecationWarning):
```
Comment (sgillies) [inline tests/test_matrix.py]: Test for profile name deprecation warning.

**Comment by sgillies** [comment]
Now it's ready. Thanks for the review @perrygeo .

**Pull request merged**